File: api/track-view.py
import json
import os
import uuid
from datetime import datetime
from http.server import BaseHTTPRequestHandler
from urllib.parse import quote
from urllib.request import Request, urlopen

KV_URL = os.environ.get('KV_REST_API_URL')
KV_TOKEN = os.environ.get('KV_REST_API_TOKEN')

def kv_call(path):
    if not KV_URL or not KV_TOKEN:
        raise RuntimeError("KV not configured")
    req = Request(f"{KV_URL}{path}")
    req.add_header("Authorization", f"Bearer {KV_TOKEN}")
    with urlopen(req) as resp:
        return json.loads(resp.read().decode())

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            content_length = int(self.headers['Content-Length'])
            data = json.loads(self.rfile.read(content_length).decode('utf-8'))

            page = data.get('page')
            session_id = data.get('sessionId')
            user_agent = data.get('userAgent', '')
            timestamp = data.get('timestamp', datetime.now().isoformat())

            if not page or not session_id:
                self.send_response(400)
                self.send_header('Content-type', 'application/json')
                self.end_headers()
                self.wfile.write(json.dumps({'error': 'Missing page or sessionId'}).encode())
                return

            view = {
                'id': str(uuid.uuid4()),
                'page': page,
                'sessionId': session_id,
                'userAgent': user_agent,
                'ip': (self.headers.get('X-Forwarded-For', '') or self.headers.get('X-Real-IP', '') or 'unknown').split(',')[0].strip(),
                'timestamp': timestamp,
                'date': datetime.fromisoformat(timestamp.replace('Z', '+00:00')).strftime('%Y-%m-%d')
            }

            # Persist to KV: LPUSH + LTRIM to keep last 10k
            payload = quote(json.dumps(view))
            kv_call(f"/LPUSH/views/{payload}")
            kv_call("/LTRIM/views/0/9999")

            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            self.end_headers()
            self.wfile.write(json.dumps({'success': True, 'message': 'View tracked successfully', 'viewId': view['id']}).encode())

        except Exception as e:
            print(f"Error tracking view: {str(e)}")
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*')
            self.end_headers()
            self.wfile.write(json.dumps({'error': 'Internal server error'}).encode())

    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

File: api/track-view.js
// Vercel API endpoint: /api/track-view
// This endpoint receives view data and stores it in a JSON file

import fs from 'fs';
import path from 'path';

export default async function handler(req, res) {
  // Only allow POST requests
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    const { page, sessionId, userAgent, timestamp } = req.body;

    // Validate required fields
    if (!page || !sessionId) {
      return res.status(400).json({ error: 'Missing required fields' });
    }

    // Get client IP (for unique user tracking)
    const clientIP = req.headers['x-forwarded-for'] || 
                    req.headers['x-real-ip'] || 
                    req.connection.remoteAddress || 
                    'unknown';

    // Create view record
    const viewRecord = {
      id: `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
      page,
      sessionId,
      userAgent: userAgent || 'unknown',
      ip: clientIP,
      timestamp: timestamp || new Date().toISOString(),
      date: new Date().toISOString().split('T')[0] // YYYY-MM-DD format
    };

    // Read existing data
    const dataPath = path.join(process.cwd(), 'data', 'views.json');
    let views = [];

    try {
      if (fs.existsSync(dataPath)) {
        const fileContent = fs.readFileSync(dataPath, 'utf8');
        views = JSON.parse(fileContent);
      }
    } catch (error) {
      console.error('Error reading views file:', error);
      views = [];
    }

    // Add new view
    views.push(viewRecord);

    // Keep only last 10,000 records to prevent file from getting too large
    if (views.length > 10000) {
      views = views.slice(-10000);
    }

    // Write back to file
    try {
      // Ensure data directory exists
      const dataDir = path.dirname(dataPath);
      if (!fs.existsSync(dataDir)) {
        fs.mkdirSync(dataDir, { recursive: true });
      }

      fs.writeFileSync(dataPath, JSON.stringify(views, null, 2));
    } catch (error) {
      console.error('Error writing views file:', error);
      return res.status(500).json({ error: 'Failed to save view data' });
    }

    // Return success
    res.status(200).json({ 
      success: true, 
      message: 'View tracked successfully',
      viewId: viewRecord.id
    });

  } catch (error) {
    console.error('Error tracking view:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
}
